### PR TITLE
Apply algorithm in scheduler by feature gates.

### PR DESCRIPTION
**Pull request by k82cn**
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: part of #42001

**Release note**:
```release-note
Apply algorithm in scheduler by feature gates.
```

### CODE REVIEW

**Comment by k82cn** [comment]
/cc @gmarek 

Working on the integration test for TaintNodeByCondition :).

File: test/integration/scheduler/taint_test.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+	// Start NodeController for taint.
+	nc, err := node.NewNodeController(
+		informers.Core().V1().Pods(),
+		informers.Core().V1().Nodes(),
+		informers.Extensions().V1beta1().DaemonSets(),
+		nil,
+		clientset,
+		time.Second,
+		100,
```
Comment (gmarek) [inline test/integration/scheduler/taint_test.go]: either use named variables for all those values, or add comments by them explaining which is what.

**Comment by k82cn** [comment]
@gmarek, I think it's ready for review :).

There's one issue I'd like to highlight here: the PodToleratesNodeTaints is enabled in `init()` (`defaultPredicates`) which
can not read FeaturesGate from args; so I introduces `algorithmprovider.Install()` to register predicates/priority explicitly.

/cc @bsalamat for such enhancement.

File: plugin/cmd/kube-scheduler/app/server.go [hunk 2]
```diff
@@ -77,6 +77,9 @@ func Run(s *options.SchedulerServer) error {
 	// cache only non-terminal pods
 	podInformer := factory.NewPodInformer(kubecli, 0)
 
+	// Install algorithm provider and algorithms before scheduling.
+	algorithmprovider.Install()
```
Comment (gmarek) [inline plugin/cmd/kube-scheduler/app/server.go]: I believe the same needs to be done in `./app/configurator.go`. Maybe... This change can actually break some tooling, but I'm not sure how much we care. @kubernetes/sig-scheduling-pr-reviews

**Comment by gmarek** [comment]
This should be split into code changes and test, to make the review easier.

**Comment by k82cn** [comment]
ok, let me split this into two PR :).

**Comment by k82cn** [comment]
Updated commits to only handle enhancement on scheduler.

@kubernetes/sig-scheduling-pr-reviews

**Comment by gmarek** [comment]
There's an issue with this being a rather big change this late in release. It's needed to write test for taint-for-condition. @davidopp

**Comment by k82cn** [comment]
The integration test is here: #53184 :).

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 3]
```diff
@@ -190,19 +191,33 @@ func defaultPredicates() sets.String {
 			},
 		),
 	)
+}
+
+// ApplyFeatures applys algorithm by feature gates.
+func ApplyFeatures() {
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: I am not sure about the name of this function and I think separating this from `defaultPredicates` reduces readability. Now, one could get wrong impression about the set of default predicates by looking at `defaultPredicates` function.

File: plugin/pkg/scheduler/algorithmprovider/plugins_test.go [hunk 4]
```diff
@@ -63,3 +64,46 @@ func TestAlgorithmProviders(t *testing.T) {
...
+			break
+		}
+
+		if !p.FitPredicateKeys.Has("PodToleratesNodeTaints") {
+			t.Errorf("Failed to find predicate: 'PodToleratesNodeTaints'")
+			break
+		}
+	}
+
+	// Appy features for algorithm providers.
```
[the remark this answers is not in the document]
Comment (k82cn) [reply plugin/pkg/scheduler/algorithmprovider/plugins_test.go]: ops :), done.

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 3 above)
Comment (k82cn) [reply plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: Fair point :). Here's the option I used to try:

1. Register default predicates in `init()`, and use `ApplyFeatures` to update/apply predicates by features gate
1. Move the logic in `init()` to a new func, named `Install()`; and call `Install()` in both `init()` and `app.Run`
1. Similar to /2, but call `Install()` explicitly if necessary
1. Update getFitPredicateFunctions to return predicates by feature gate

The PR follows option 1 :). The second option requires `Install()` can be re-run, and the result dependent on feature gates,
the function user maybe confused with the result. The drawback of option 3 is risk, if we missed a `Install()`, some feature
may not work.

Comment (bsalamat) [reply plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: I am a bit confused. I thought 'defaultPredicates()' which was being called from `init()` was already registering the right predicate functions based on feature gates. So, why is the specific change that this PR make in registering predicates?

Comment (k82cn) [reply plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: The `init()` will be executed when package imported; the feature gate is not updated by args at that time.

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 5]
```diff
@@ -190,19 +191,33 @@ func defaultPredicates() sets.String {
 			},
 		),
 	)
+}
+
+// ApplyFeatures applys algorithm by feature gates.
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: s/applys/applies/

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 3 above)
Comment (bsalamat) [reply plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: I see. I would call this function something like "ApplyFeatureGates" or something like that to make it more clear what it does.

File: plugin/pkg/scheduler/algorithmprovider/plugins_test.go [hunk 6]
```diff
@@ -63,3 +64,46 @@ func TestAlgorithmProviders(t *testing.T) {
...
+		if !p.FitPredicateKeys.Has("PodToleratesNodeTaints") {
+			t.Errorf("Failed to find predicate: 'PodToleratesNodeTaints'")
+			break
+		}
+	}
+
+	// Apply features for algorithm providers.
+	utilfeature.DefaultFeatureGate.Set("TaintNodesByCondition=True")
+
+	ApplyFeatures()
```
Comment (kevin-wangzefeng) [inline plugin/pkg/scheduler/algorithmprovider/plugins_test.go]: s/ApplyFeatures/ApplyFeatureGates/

**Comment by bsalamat** [comment]
/lgtm

Thanks, @k82cn!

**Comment by k82cn** [comment]
@gmarek , do nodeController part ok to you?

**Comment by mindprince** [comment]
@k82cn This caused GKE GPU tests and GKE Device plugin tests to fail:

Diff between passing and failing runs:
https://github.com/kubernetes/kubernetes/compare/924bd563b...4c115a862

GKE Device Plugin Tests:
https://k8s-testgrid.appspot.com/wg-resource-management#gke-device-plugin-gpu
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gke-device-plugin-gpu/95

GKE GPU Tests:
https://k8s-testgrid.appspot.com/wg-resource-management#gke-gpu
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gke-gpu/1453

/cc @vishh @jiayingz

**Comment by k82cn** [comment]
wow, let me handle that :).

**Pull request merged**